drivers: video: hm01b0 - more formats

KurtE opened the pull request:
Now that the @NinoSc PR has been merged:

Adding some of the missing pieces to the HM01b0 video class.

Note: not maybe necessary, but changed the VIDEO define to match all of the other video drivers that I have tried:
That is CONFIG_VIDEO_HIMAX_HM01B0 to
CONFIG_VIDEO_HM01B0

Also keep wondering if the default data-bits should be 8 as to be more likely compatible with more camera drivers.  But
left it as 1.

Added PWDN and RESET pins, to yaml, plus code.

Added a few more format sizes: 326x326 to 326x244 match the spec.  Verified again from earlier 
tests, and Logic Analyzer.

I have also added some of the other components that are in other related drivers, like the
ctrl for HFLIP/VFlip

Added 4 bit mode using the format @avolmat-st suggested in the related issue.

Added format for cameras with Bayer filter.

Questions are: 
Should I remove the current resolutions like: 320x320 as the camera does not actually return that?

The images are not as good as I would like or saw from the cameras on other setups.  I believe a lot of
this is the other setups (Arducam, Linux, Teensy) do a lot settings of the registers than this currently does.
Things like Auto Exposure and the like.

Main question is: what granularity of change would you prefer to see? 
Continue on to refine the images?  Or incremental steps?  Should I squash all of these?
Do you agree with the cahnge from CONFIG_VIDEO_HIMAX_HM01B0 -> CONFIG_VIDEO_HM01B0 or should I revert it?

Thanks
Kurt

josuah commented on drivers/video/hm01b0.c:
```diff
@@ -261,10 +441,84 @@ static bool hm01b0_check_connection(const struct device *dev)
+	const struct hm01b0_config __maybe_unused *config = dev->config;
 	int ret;
 
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(pwdn_gpios) || DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
```
You do not need that "double-`#ifdef`", or at least not at this place, proposition to move it below to simplify a bit.

josuah commented on drivers/video/hm01b0.c:
```diff
@@ -261,10 +274,84 @@ static bool hm01b0_check_connection(const struct device *dev)
+	LOG_INF("hm01b0_init check connection");
 	if (!hm01b0_check_connection(dev)) {
 		LOG_ERR("%s is not ready", dev->name);
 		return -ENODEV;
```
If I understand correctly, here we have first try to configure reset/pwdn pin as needed, and then try to toggle them repeatedly a configurable number of times?

It seems like the latter is all what need to be supported as it would act as both?

```suggestion
	for (int i = 0; i < HM01B0_INIT_ATTEMPTS; i++) {
#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
		if (config->reset.port != NULL) {
			if (!gpio_is_ready_dt(&config->reset)) {
				return -ENODEV;
			}
			ret = gpio_pin_configure_dt(&config->reset, GPIO_OUTPUT_ACTIVE);
			if (ret < 0) {
				LOG_ERR("Could not set reset pin: %d", ret);
				return ret;
			}
		}
#endif

#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
		/* Reset camera module */
		if (config->reset.port != NULL) {
			if (!gpio_is_ready_dt(&config->reset)) {
				return -ENODEV;
			}
			ret = gpio_pin_configure_dt(&config->reset, GPIO_OUTPUT_ACTIVE);
			if (ret < 0) {
				LOG_ERR("Could not set reset pin: %d", ret);
				return ret;
			}
		}
#endif
	}

	if (!hm01b0_check_connection(dev)) {
		LOG_ERR("%s is not ready", dev->name);
		return -ENODEV;
	}
```

Did I understand the goal corectly?

KurtE replied:
The problem with your possible replacement code, is that it  potentially can try all 4 combination of
pin setting as I mentioned, but if the first one does not succeed it will return (-ENODEV).

The current code Will fail in the call hm01b0_try_reset_pwdn_pinsm which does like hm01b0_check_connection,
except it will not bail at this point but continue to the next combination and try again.

josuah commented on drivers/video/Kconfig.hm01b0:
```diff
@@ -1,6 +1,6 @@
 # Copyright The Zephyr Project Contributors
 # SPDX-License-Identifier: Apache-2.0
-config VIDEO_HIMAX_HM01B0
+config VIDEO_HM01B0
```
Good catch thank you!

You might remove the `not maybe necessary,` from the commit if wanting a bit more canonical style, where discussion about "is this the correct thing to do" is left outside the commit message.

Likewise "so will test this out" is perfect for the PR discussion thread, but in the commit message, best to only have description of the changes... The same content as a big `/* code comment */` describing the patch.

In an email, there would be first a textual comment in prose *before* the commit messages then the patches with their short commit message.

Just a matter of conventions.

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -1338,6 +1338,15 @@ int video_estimate_fmt_size(struct video_format *fmt);
+/**
+ * @code{.unparsed}
+ *   0          1          2          3          3 2 1 0
+ * | 0000Yyyy | 0000Yyyy | 0000Yyyy | 0000Yyyy | 0000Yyyy | ...
```
Is this really Y4?
You probably meant this:
```
    0          1          2          3
  | 0000Yyyy | 0000Yyyy | 0000Yyyy | 0000Yyyy | ...
```

If we consider that Y10P is this:
```
  0          1          2          3          3 2 1 0
| Yyyyyyyy | Yyyyyyyy | Yyyyyyyy | Yyyyyyyy | yyyyyyyy | ...
```

And that `Y10` is this:

```
  0                   1                   2                   3
| yyyyyyyy 000000Yy | yyyyyyyy 000000Yy | yyyyyyyy 000000Yy | yyyyyyyy 000000Yy | ...
| yyyyyyyy 000000Yy | yyyyyyyy 000000Yy | yyyyyyyy 000000Yy | yyyyyyyy 000000Yy | ...
```

Then it makes sense that `Y4P` is this:

```
    0   1      2   3
  | YyyyYyyy | YyyyYyyy | ...
```

And that `Y4` is this:

```
    0          1          2          3
  | 0000Yyyy | 0000Yyyy | 0000Yyyy | 0000Yyyy | ...
```

And I agree that it would be what a video receiver gets if having 4 pins hooked in out of 8 pins.

Should be reasonable to integrate!

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -1338,6 +1338,15 @@ int video_estimate_fmt_size(struct video_format *fmt);
+ * | 0000Yyyy | 0000Yyyy | 0000Yyyy | 0000Yyyy | 0000Yyyy | ...
+ * @endcode
+ */
+#define VIDEO_PIX_FMT_Y4      VIDEO_FOURCC('Y', '0', '4', ' ') /*  4  Greyscale     */
```
Linux adds comment at the end of the pixel format, but they are not rendered in Doxygen so can be removed from Zephyr. Also you may remove the spaces after `Y4`...

```
#define VIDEO_PIX_FMT_Y4 VIDEO_FOURCC('Y', '0', '4', ' ')
```

josuah commented on drivers/video/hm01b0.c:
```diff
@@ -284,6 +285,73 @@ static int hm01b0_soft_reset(const struct device *dev)
+	} else {
+		/* Set to the max possible FPS at this resolution. */
+		osc_div = 3;
+	}
```
This is applying a `ceil()` to the frame interval from the user, it's also possible to use `void video_closest_frmival()`
for getting the closest match, and then you can check with `==`.

https://github.com/nxp-upstream/zephyr/blob/40f07a0532ed1ddabf33180eb2be0618bc19e832/include/zephyr/drivers/video.h#L933-L950

However, this requires also supporting `video_enum_frmival()`, which is mandatory for supporting frame intervals.

Given you are one of the first to implement this API on an image sensor there are not a lot of examples for how to do it. You could try, like suggested in another PR by Phi Bang, without ` = 60/30/15`:

```c
enum {
		HM01B0_60_FPS,
		HM01B0_30_FPS,
		HM01B0_15_FPS,
};
```

Then here:

```suggestion
	struct video_frmival_enum fie = {.format = &data->fmt, .discrete = *frmival};

	video_closest_frmival(dev, &fie);
	
	switch (fie.index) {
	case HM01B0_60_FPS:
		osc_div = highres ? 0x01 : 0x00;
		break;
	case HM01B0_30_FPS:
		osc_div = highres ? 0x02 : 0x01;
		break;
	case HM01B0_15_FPS:
		osc_div = highres ? 0x03 : 0x02;
		break;
	default:
		LOG_ERR("Unsupported frame interval value");
		return -ENOTSUP;
	}
```

josuah replied:
And to implement `hm01b0_enum_frmival()` I think this might work:

```c
int hm01b0_enum_frmival(const struct device *dev, struct video_frmival_enum *fie)
{
	switch (fie->index) {
	case HM01B0_60_FPS:
		fie->discrete.denominator = 60;
		break;
	case HM01B0_30_FPS:
		fie->discrete.denominator = 30;
		break;
	case HM01B0_15_FPS:
		fie->discrete.denominator = 15;
		break;
	default:
		return -EINVAL;
	}

	fie->type = VIDEO_FRMIVAL_TYPE_DISCRETE;
	fie->discrete.numerator = 1;

	return 0;
}
```

mjs513 replied:
@josuah 
@KurtE  - pointed out that you suggested this change, since I have the same in the HM0360 will update that after I fix up my pr issues.

Thanks guys

KurtE replied:
done, I also added 4th for lower speed.

josuah reviewed (commented):
Thanks for these improvements!

Here are a few suggestions to try to reduce the code size and some help for implementing frame rate control.

Let me know if you have questions!

KurtE commented:
I just pushed up, changes that initializes a lot more of the registers using tables that was common
to several other setups, like Arduino MBED, Teensy, Arducam.

Getting better images and it fixed that we are getting 326 is now 324 as expected.

There are issues on how many rows are generated, which I am working through.  May have been the same on the other
platforms as they only take one frame at a time, and then sync back up start of frame.

But with these updated changes, My Zephyr example sketch(sample) where I added quick and dirty BAYER to rgb565 output
is showing a lot more data now:  It is scrolling on each frame, which is indicative of the issue. 

Going to mark back to in progress, but pushed up to let other try if the desire.
<img width="1092" height="632" alt="image" src="https://github.com/user-attachments/assets/24f0e753-63a6-4491-938e-9a8fd958ca11" />

KurtE commented:
@josuah @mjs513 I made the changes to use the:
```
		ret = video_modify_cci_reg(&config->i2c, HM01B0_CCI_IMG_ORIENTATION,
					   HM01B0_ORIENTATION_HMIRROR,
					   FIELD_PREP(HM01B0_ORIENTATION_HMIRROR,
						      ctrls->hflip.val));
```
as was done in #94904 although I did modify the order of the arguments.
Using my own zephyr test app, I tried 3 out of the 4 combinations and they appeared to work.

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -1377,6 +1338,15 @@ int video_transfer_buffer(const struct device *src, const struct device *sink,
+ * @endcode
+ */
+#define VIDEO_PIX_FMT_Y4 VIDEO_FOURCC('Y', '0', '4', ' ')
+
```
For this, it would be best to have a separate commit before the driver as it is not specific to this image sensor (at least... in theory).

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -1377,6 +1338,15 @@ int video_transfer_buffer(const struct device *src, const struct device *sink,
+/**
+ * @code{.unparsed}
+ *   0                   1                   2
+ * | 0000yyyy 0000Yyyy | 0000yyyy 0000Yyyy | 0000yyyy 0000Yyyy | ...
```
Some small improvement would be to use 4 pixels: all formats show how to encode 4 pixels on purpose to give an estimation of how large the data is in RAM at first glance.

```suggestion
 *   0                   1                   2                   3
 * | 0000yyyy 0000Yyyy | 0000yyyy 0000Yyyy | 0000yyyy 0000Yyyy | 0000yyyy 0000Yyyy | ...
```

Thanks for adding the doc!

KurtE replied:
done

josuah commented on drivers/video/hm01b0.c:
```diff
@@ -261,10 +606,84 @@ static bool hm01b0_check_connection(const struct device *dev)
+		}
+		retry_count--;
+	}
+	return false;
```
But that might not need to be changed if the approach is changed (see comment below).

josuah reviewed (commented):
Thanks for the answers and the modifications!

I've left some feedback about  the reset, hoping it is possible to simplify this as if there is effectively something that needs to be addressed possibly all other sensors could face the same thing.

Let me know what you think below...

josuah commented on drivers/video/hm01b0.c:
```diff
@@ -108,6 +277,14 @@ static const struct video_format_cap hm01b0_fmts[] = {
+	HM01B0_VIDEO_FORMAT_CAP(324, 244, VIDEO_PIX_FMT_Y4),
+	HM01B0_VIDEO_FORMAT_CAP(324, 324, VIDEO_PIX_FMT_Y4),
+	HM01B0_VIDEO_FORMAT_CAP(324, 244, VIDEO_PIX_FMT_SBGGR8),
+	HM01B0_VIDEO_FORMAT_CAP(324, 324, VIDEO_PIX_FMT_SBGGR8),
```
Now that we have a little bit more formats, to help making sure that the definition is in-sync with the enum, how about this?
```suggestion
	[RESOLUTION_160x120] = HM01B0_VIDEO_FORMAT_CAP(160, 120, VIDEO_PIX_FMT_GREY),
	[RESOLUTION_320x240] = HM01B0_VIDEO_FORMAT_CAP(320, 240, VIDEO_PIX_FMT_GREY),
	[RESOLUTION_320x320] = HM01B0_VIDEO_FORMAT_CAP(320, 320, VIDEO_PIX_FMT_GREY),
	[RESOLUTION_164x122] = HM01B0_VIDEO_FORMAT_CAP(164, 122, VIDEO_PIX_FMT_GREY),
	[RESOLUTION_324x244] = HM01B0_VIDEO_FORMAT_CAP(324, 244, VIDEO_PIX_FMT_GREY),
	[RESOLUTION_324x324] = HM01B0_VIDEO_FORMAT_CAP(324, 324, VIDEO_PIX_FMT_GREY),
	[RESOLUTION_164x122_Y4] = HM01B0_VIDEO_FORMAT_CAP(164, 122, VIDEO_PIX_FMT_Y4),
	[RESOLUTION_324x244_Y4] = HM01B0_VIDEO_FORMAT_CAP(324, 244, VIDEO_PIX_FMT_Y4),
	[RESOLUTION_324x324_Y4] = HM01B0_VIDEO_FORMAT_CAP(324, 324, VIDEO_PIX_FMT_Y4),
	[RESOLUTION_324x244_BAYER] = HM01B0_VIDEO_FORMAT_CAP(324, 244, VIDEO_PIX_FMT_SBGGR8),
	[RESOLUTION_324x324_BAYER] = HM01B0_VIDEO_FORMAT_CAP(324, 324, VIDEO_PIX_FMT_SBGGR8),
```

Not super super pretty... hoepfully future APIs based on `_enum_pixelformats()` and `_enum_resolutions()` allow to have a single table rather than multiple to keep in sync.

non-blocking, just proposal for a small improvement.

josuah reviewed (commented):
Some small extra...

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -239,11 +413,166 @@ static int hm01b0_soft_reset(const struct device *dev)
+		CODE_UNREACHABLE;
+	}
+
+	osc_div |= 0x8;
```
I don't know the meaning of the 0x08, but wouldn't it be simplier to have osc_div initialized to 0x08 at the beginning and have |= within each case of the switch case above ?

KurtE replied:
To me this feels like a wash either way: One way we or on one value at the end, the other way we do |= on the
4 current assignment statements.  Can do if you want.

|= 0x8;

This is:
<img width="744" height="232" alt="image" src="https://github.com/user-attachments/assets/6b95193f-7dd6-4cda-8d48-ae46b7de1922" />

Could do a define for it.  Probably not all of them.

avolmat-st replied:
Yes, seems better to have some define here for the value you are using.

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -239,11 +413,166 @@ static int hm01b0_soft_reset(const struct device *dev)
+
+	ret = video_init_ctrl(&ctrls->hflip, dev, VIDEO_CID_HFLIP,
+			      (struct video_ctrl_range){.min = 0, .max = 1, .step = 1, .def = 0});
+	if (ret) {
```
if (ret < 0) {

KurtE replied:
done will upload

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -263,8 +592,39 @@ static bool hm01b0_check_connection(const struct device *dev)
+#endif
+#endif
+
+	LOG_INF("hm01b0_init check connection");
```
Is this really needed as a LOG_INF ?

avolmat-st commented on include/zephyr/drivers/video.h:
```diff
@@ -1338,6 +1338,15 @@ int video_estimate_fmt_size(struct video_format *fmt);
  */
 #define VIDEO_PIX_FMT_GREY VIDEO_FOURCC('G', 'R', 'E', 'Y')
 
+
```
Could you keep the video subsystem non driver modifications in a separate commit ?

avolmat-st reviewed (changes requested):
Thanks @KurtE. I understand that you keep an eye on the HM0360 PR as well so I didn't put here some related comments I made on that PR as well. For example about the usage of the hm01b0_get_fmt instead of using directly drv_data->fmt. Could you take care of those points as well ?

KurtE commented on drivers/video/hm01b0.c:
```diff
@@ -26,71 +28,235 @@ LOG_MODULE_REGISTER(hm01b0, CONFIG_VIDEO_LOG_LEVEL);
 #define HM01B0_REG8(addr)             ((addr) | VIDEO_REG_ADDR16_DATA8)
 #define HM01B0_REG16(addr)            ((addr) | VIDEO_REG_ADDR16_DATA16_BE)
 #define HM01B0_CCI_ID                 HM01B0_REG16(0x0000)
+#define HM01B0_CCI_IMG_ORIENTATION    HM01B0_REG8(0x0101)
```
done

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -26,71 +28,235 @@ LOG_MODULE_REGISTER(hm01b0, CONFIG_VIDEO_LOG_LEVEL);
 
 struct hm01b0_config {
 	const struct i2c_dt_spec i2c;
 	const uint8_t data_bits;
```
Yes this is not something done in this commit I agree. This can be removed in another comment if you wish or later on in another PR

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -239,11 +413,166 @@ static int hm01b0_soft_reset(const struct device *dev)
+	struct video_frmival_enum fie = {.format = &drv_data->fmt, .discrete = *frmival};
+
+	/* get_fmt always returns 0 */
+	hm01b0_get_fmt(dev, &fmt);
```
What about directly relying on the drv_data->fmt structure here instead of calling hm01b0_get_fmt which basically is simply going to do a copy of the drv_data->fmt into the local fmt variable.
Since drv_data->fmt is already accessed few lines above, what about having this harmonized. This also remove the need to have struct video_format fmt.

KurtE replied:
done

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -239,11 +413,166 @@ static int hm01b0_soft_reset(const struct device *dev)
+	video_closest_frmival(dev, &fie);
+
+	/*
+	 * Note: in highres mode max FPS is 45 else 60
```
I missed that comment before. If this is known that in VGA mode this is only 45FPS and not 60FPS, why not making this visible to the application by introducing a 45FPS.
In case of < VGA, this means 8 / 15 / 30 / 60 can be set & enumerated and in VGA only 8 / 15 / 30 / 45 (or is it all shifted ?)

KurtE replied:
@mjs513 @josuah @avolmat-st - I wondered that at one point.  
I believe that several other drivers ignored it as to make things feel simpler.

I would think to do this right might be to: logically have two enums or could combine them.
```
enum {
	HM01B0_60_FPS,
	HM01B0_30_FPS,
	HM01B0_15_FPS,
	HM01B0_8_FPS,
};
```
and
```
enum {
	HM01B0_45_FPS,
	HM01B0_22_FPS,
	HM01B0_11_FPS,
	HM01B0_5_FPS,
};
```
The function: hm01b0_set_frmival would need to be updated to handle these additional values.

Likewise the get_frmival might be OK, or might need to be updated. As for example:
if you ask for 45 in QQVGA mode it might only do 30 or would it 60 not sure, what closest will do.

So then back to get, should it return the value passed in, or a computed actual.  will play some

KurtE replied:
I like.  I am wondering if it would be better to cheat:
```
enum {
	HM01B0_5_OR_8_FPS,
	HM01B0_11_OR_15_FPS,
	HM01B0_22_OR_30_FPS,
	HM01B0_45_OR_60_FPS,
};
```
As: you might be able to use the index returned to be the partial register value:
<img width="745" height="235" alt="image" src="https://github.com/user-attachments/assets/27bce199-f806-4324-b68e-bf1b2f75b8f6" />

Will try it and see what happens.

josuah replied:
last I remember is highest FPS first, so could have been `3 - fie.index` for instance.

KurtE commented on drivers/video/hm01b0.c:
```diff
@@ -239,11 +413,166 @@ static int hm01b0_soft_reset(const struct device *dev)
+	int ret;
+	uint32_t osc_div = 0;
+	bool highres = false;
+	struct video_frmival_enum fie = {.format = &drv_data->fmt, .discrete = *frmival};
```
done

KurtE committed:
drivers: video: add VIDEO_PIX_FMT_Y4

Add the format to video.h
```
/**
 * @code{.unparsed}
 *   0                   1                   2                   3
 * | 0000yyyy 0000Yyyy | 0000yyyy 0000Yyyy | 0000yyyy 0000Yyyy | ...
 * @endcode
 */
```

Signed-off-by: Kurt Eckhardt <kurte@rockisland.com>

KurtE committed:
drivers: video: hm01b0 - add missing pieces

Adding some of the missing pieces to the HM01b0 video class.

changed the VIDEO define to match all of the other video drivers
that I have tried:
That is CONFIG_VIDEO_HIMAX_HM01B0 to CONFIG_VIDEO_HM01B0

Added a few more format sizes: 324x324 to 324x244 164x122
match the spec.

Added BAYER format for the color camera (VIDEO_PIX_FMT_SBGGR8)
Only for Full, and QVGA, as cameras with BAYER filter do
not support the binning mode.

Added 4 data bit support, using new format: VIDEO_PIX_FMT_Y4

Support for hflip/vflip controls.

Added the frmival set/get/enum functions, to allow us better
control of how many frames are generated per second.

Used the reset register enumeration similar to other implementions,
such as Arduino library, Teensy Library, and OpenME

Signed-off-by: Kurt Eckhardt <kurte@rockisland.com>

avolmat-st commented:
Thanks @KurtE for the update. This look good to me. However could you clarify the point regarding 45fps ?

cf https://github.com/zephyrproject-rtos/zephyr/pull/98154#discussion_r2497748539

josuah commented on drivers/video/hm01b0.c:
```diff
@@ -239,32 +415,231 @@ static int hm01b0_soft_reset(const struct device *dev)
+		break;
+	default:
+		return -EINVAL;
+	}
```
As seen here https://github.com/zephyrproject-rtos/zephyr/pull/98154/files#r2500836963 the frame rate depends on the resolution

This is the reason why `video_frmival_enum` contains a `fie->fmt` field, so that the framerate returned can be made specific to the format.

For instance:

```suggestion
	bool highres = (drv_data->fmt.width >= 320 && drv_data->fmt.height >= 320);

	switch (fie->index) {
	case HM01B0_45_OR_60_FPS:
		fie->discrete.denominator = highres ? 45 : 60;
		break;
	case HM01B0_22_OR_30_FPS:
		fie->discrete.denominator = highres ? 22 : 30;
		break;
	case HM01B0_11_OR_15_FPS:
		fie->discrete.denominator = highres ? 11 : 15;
		break;
	case HM01B0_5_OR_8_FPS:
		fie->discrete.denominator = highres ? 5 : 8;
		break;
	default:
		return -EINVAL;
	}
```

mjs513 replied:
@KurtE 
Was testing this out on the 360 and found that at least set_frmival was correct in terms of configuring the framerate reg values.  

However, when I was testing on the GIga through the arduino ide I was not seeing much change in fps. 
For 10 fps seeing about 26 fps
For 15 seeing about 25 fps
For 30 seeing about 26 fps.
At 60 seeing about 12 fps which is strange to me

So looks like no matter what something else is off

josuah replied:
It does not look like an issue in this driver, thank you for investigating.

I will try to keep an eye on this and find time to reproduce.

josuah reviewed (commented):
Some complement regarding frmival and highres mode.
Thanks again for the progress on this sensor!

josuah commented:
CI tests complain about a documentation change for something that was not modified by this PR.
This happens from time to time when some new syntax rules are enforced.
To solve this one, a `git rebase` to the latest `main` branch will be enough to solve the CI issue.

Not a problem with what is submitted...
Thanks!

KurtE commented on drivers/video/hm01b0.c:
```diff
@@ -239,32 +415,231 @@ static int hm01b0_soft_reset(const struct device *dev)
+
+	if (drv_data->fmt.width == 320 && drv_data->fmt.height == 320) {
+		highres = true;
+	}
```
Thanks, will do.
I changed it slightly as I believe it is HIGHRES if WIDTH >= 320 as it mentioned VGA.

KurtE replied:
done

KurtE commented:
> From an hardware point of view, data coming from the sensor are same (same pixel clock), so if the use-case works fine at 45fps, it should also work fine even if the DCMI later on introduce frame dropping. In both cases, data enter the DCMI are same pace and also comes out of the DCMI at same pace as well. Simply in case of frame dropping there are times when nothing comes out of the DCMI since this is fully dropped internally by the DCMI.

I hear what you are saying, however I am pretty sure in the current configuration, the camera does not work properly at 45fps, which is why a user may have asked for 20.   Or maybe it uses less energy and they are running on a battery,  Or maybe 
image quality is different (no idea), ...

As I currently for this camera have it defined to run in 4 bit mode and I am running with clock of 6mhz
<img width="404" height="367" alt="image" src="https://github.com/user-attachments/assets/9aeb194d-4acb-46cb-9016-50ec7ff19d9f" />

To get 45 fps, I would need to run at 12mhz and maybe there could be issues with running it at 12Mhz (have not checked it in
awhile.  Things like the signal going through the Portenta to the carrier board, to the camera connector and then to the
camera module.

avolmat-st commented:
Thanks for the clarification. So, in fact the problem is that the driver is reporting that the sensor can produce 45fps, while in fact it cannot do so, right ?
So, in current situation the driver should just not advertised framerate that it cannot produce.
That is the first fix to be done I think.

One additionnal fix could also be done to have the driver smart enough to advertise possible framerate based on its input quartz frequency. This is also possible by having an input clock referenced to the driver and described in the device-tree.
Then, depending on the HW configuration of the sensor (quartz of 6MHz or 12MHz), the sensor driver will advertize possible framerate as well as maybe adjust as well some PLL settings if they (probably) differ depending on the input frequency of the sensor.
An example for this can be found within the imx335.c sensor, you might want to have a look at imx335_set_input_clk function, as well as the input_clk variable part of the ctrl structure and set by the init function.
And in order to indicate to the sensor driver the input clock, this is done within the device tree, again, the example you can look at is in boards/shields/st_b_cams_imx_mb1854/st_b_cams_imx_mb1854.overlay

```c
        imx335_input_clock: imx335-input-clock {
                compatible = "fixed-clock";
                clock-frequency = <24000000>;
                #clock-cells = <0>;
        };
```

and

```c
        imx335: camera@1a {
                compatible = "sony,imx335";
                clocks = <&imx335_input_clock>;
                reg = <0x1a>;
                reset-gpios = <&csi_connector CSI_IO0 GPIO_ACTIVE_LOW>;

                port {
                        imx335_ep_out: endpoint {
                                remote-endpoint-label = "csi_ep_in";
                                bus-type = <VIDEO_BUS_TYPE_CSI2_DPHY>;
                                data-lanes = <1 2>;
                        };
                };
        };
```

This is not mandatory to do all this now, you can stick to 6MHz only input but in such case update the possible framerates since I understand that at least 45fps is not achievable if the input frequency is 6MHz.

KurtE commented:
@avolmat-st - will look into more advanced ways to control it.  Maybe for the next round of changes.

Note: It is unclear to me, which hardware may potentially work at the highest speeds.  As I mentioned, it might be
very temperamental, like wire lengths and timings. 

So far kept it simple, if you are running in 4 bit mode, which you do if you are using Arducam/Arduino camera, it restricts
out the 45/60 index:

Other options:  Add option to device tree for the camera that limits the frame rates. i.e. add to the dts himax,hm01b0.yaml file
option like: max-framerate 
which gives the limit.  Could be value/values depending on if you wish to distinguish between high-res or not. 

Or provide an easy way for the end app to say, really give me the nearest one to the speed I ask for.  If it does not work
I will choose a different one.  As I have a handle to the actual camera object:

From my test app
```
#ifdef FRAME_RATE
	struct video_frmival frmival;
	struct video_frmival_enum fie;

	printk("Request Frame Rate: %u\n", FRAME_RATE);

	frmival.denominator = FRAME_RATE;
	frmival.numerator = 1;

	fie.format = &fmt;
	fie.discrete = frmival;
	fie.type = VIDEO_FRMIVAL_TYPE_DISCRETE;
	video_closest_frmival(video_dev, &fie);
	printk("Closest dev: t:%u %u %u\n", fie.type, fie.discrete.numerator, fie.discrete.denominator);

	if (video_dev_sensor) {
		fie.format = &fmt;
		fie.discrete = frmival;
		fie.type = VIDEO_FRMIVAL_TYPE_DISCRETE;
		video_closest_frmival(video_dev_sensor, &fie);
		printk("Closest sensor: t:%u %u %u\n", fie.type, fie.discrete.numerator, fie.discrete.denominator);
	}

	if (video_set_frmival(video_dev, &frmival) < 0){
		printk("ERROR: Unable to set up frame rate\n") ;
	}

	video_get_frmival(video_dev, &frmival);
	printk("Returned frame rate: %u\n", frmival.denominator);

#endif
```
Output when I requested 20fps
```
Request Frame Rate: 20
Closest dev: t:1 162000 1822500
Closest sensor: t:1 1 22
Returned frame rate: 45
```

But if I took what the sensor tells me of 22, it will give me 22. So there is an out.

I am personally wish that the DCMI scaling was optional.  That is maybe be able to turn that on or off.  As I may 
choose the lower speeds for other reasons.  But that is beyond this scope.

Sort of in a different context, if I told my car I wish to go 30 miles per hour and the red line for the engine
was 6000rpm, I would not prefer for it to choose to run in 1st gear at 6000 rpm as it thought it was the
most precise.  I know different, but ...

avolmat-st commented:
> @avolmat-st - will look into more advanced ways to control it. Maybe for the next round of changes.
> 
> Note: It is unclear to me, which hardware may potentially work at the highest speeds. As I mentioned, it might be very temperamental, like wire lengths and timings.
> 
> So far kept it simple, if you are running in 4 bit mode, which you do if you are using Arducam/Arduino camera, it restricts out the 45/60 index:
> 
> Other options: Add option to device tree for the camera that limits the frame rates. i.e. add to the dts himax,hm01b0.yaml file option like: max-framerate which gives the limit. Could be value/values depending on if you wish to distinguish between high-res or not.

To me that's not the max allowed framerate that has to be mentioned in the device tree but the max pixel clock rate. This is an idea I have in mind already since I need to limit the OV5640 pixel clock out of the sensor for some of the STM32 boards.
We already have something similar for the CSI for which we can mention the link-frequency. For CSI the next level is to be able to mention applicable link-frequencies in the DT so that driver can only advertised resolution/framerate that can be actually performed by the HW.
For DVP, this is similar, in that a receiver should be able to mention the maximum PCLK is can actually handle.
Doing so, on your setup you'd mention that you can only handle 6MHz, and thus 45/60fps won't be visible.

> 
> Or provide an easy way for the end app to say, really give me the nearest one to the speed I ask for. If it does not work I will choose a different one. As I have a handle to the actual camera object:
 
The API is here to, apply a framerate requested by the application, and if the framerate is not achievable by the HW, adjust it. It is after that up to the application to check what the lower layers have applied (by checked the applied framerate via the video_get_frmival) and if it is not happy with this, set another thing.

> I am personally wish that the DCMI scaling was optional. That is maybe be able to turn that on or off. As I may choose the lower speeds for other reasons. But that is beyond this scope.

Seems to me that only reason why DCMI frame drop would have to be disabled is because some of the framerate advertised by the sensor cannot be achieved while they are advertised.

avolmat-st reviewed (changes requested):
Could this simply be that for some reason, when the 45fps / 60fps is set, the sensor end up dropping 1/4 of the frame, and end up "generating" frame like 160x240 instead of 320x240 ? Basically it wouldn't have enough time to send a while line between each HSYNC.

If that was the case, this would I think lead to such image since, the display would be reason each line as 320 pixels, hence left part of the display being in fast each odd lines and right part the even lines.
And since the H7 DMA is configured to capture a fixed size of 320x240 * 2, (? ... well anyway, a fixed size), it end up captuing actually 2 frames and putting them into a single frame.

mjs513 commented:
@KurtE - @avolmat-st 
Since Kurt seems to be having issues with frmival I decided to test my implementation since they are really similar.  I did notice several strange things going on:
If I set framerate to 20 in the arduino camera library I am using it is shows the the libary is returnin 60 fps:
```
Requested FrameRate: 20
Returned FrameRate: 60
```
Now if I look at the debug info from the set_frmival itself things get weird.
```
_set_frmival: FrameRate selected: 60 
hm0360_set_frmival: Applying frame interval number 0
hm0360_set_frmival: OSC DIV 60: 0x1
 HIRES Selected: 0
FrameRate rounded to: 60
```
Note framerate selected should be 20 (what I input) not 60? 

Now before you say its the arduino code if I specify a framerate of 30:
```
_set_frmival: FrameRate selected: 30
hm0360_set_frmival: Applying frame interval number 1 (which is 30fps)
hm0360_set_frmival: OSC DIV 30: 0x2
FrameRate rounded to: 30
HIRES Selected: 0
```
which seems to work 10 and 15 work as well but if I select 40 for instance I would think 30 should be closest and not 60 but am seeing 60.

Also I notced that the OV5640 driver does frmival a bit different and does not use video_frmival_closest but uses a best match setup

Sorry to jump in on the HM01b0 but @KurtE tend to compare notes.

avolmat-st commented:
> @KurtE - @avolmat-st Since Kurt seems to be having issues with frmival I decided to test my implementation since they are really similar. I did notice several strange things going on: If I set framerate to 20 in the arduino camera library I am using it is shows the the libary is returnin 60 fps:
> 
> ```
> Requested FrameRate: 20
> Returned FrameRate: 60
> ```

Good catch. Actually there is a bug in the return value given back by the DCMI driver.

The DCMI driver set_frmival finish with:

```c
        /*
         * Give back the achieved frame interval achieved, ensuring to take into
         * consideration the DCMI frame control
         */
        frmival->numerator = best_sensor_frmival.numerator * best_capture_rate;
        frmival->denominator = best_sensor_frmival.denominator;

        data->capture_rate = best_capture_rate;

        return video_set_frmival(config->sensor_dev, &best_sensor_frmival);
```

Meaning that, after it tries to figure out what would be closest frame-interval, taking into consideration sensor frmival and what the DCMI IP can achieve, it end up peeking up the 60fps for the sensor (which, I must say is not the best choice here in fact since this will lead to either 30fps or 15fps depending on the frame-drop which is calculated for the DCMI) and this exact framerate could also be generated by the sensor itself in fact. So here the DCMI driver should be smarter, and picking up the sensor frmival directly instead of using it own since anyway frmival is the same).
However there is another bug which is the return frmival at the end. Here it is directly return best_sensor_frmival which this should have been best_sensor_frmival divided by the frame drop applied for the DCMI.

Do you wish to make a patch for that or do you want me to make it ?

> Now if I look at the debug info from the set_frmival itself things get weird.
> 
> ```
> _set_frmival: FrameRate selected: 60 
> hm0360_set_frmival: Applying frame interval number 0
> hm0360_set_frmival: OSC DIV 60: 0x1
>  HIRES Selected: 0
> FrameRate rounded to: 60
> ```
> 
> Note framerate selected should be 20 (what I input) not 60?
> 
> Now before you say its the arduino code if I specify a framerate of 30:
> 
> ```
> _set_frmival: FrameRate selected: 30
> hm0360_set_frmival: Applying frame interval number 1 (which is 30fps)
> hm0360_set_frmival: OSC DIV 30: 0x2
> FrameRate rounded to: 30
> HIRES Selected: 0
> ```
> 
> which seems to work 10 and 15 work as well but if I select 40 for instance I would think 30 should be closest and not 60 but am seeing 60.
> 
> Also I notced that the OV5640 driver does frmival a bit different and does not use video_frmival_closest but uses a best match setup
> 
> Sorry to jump in on the HM01b0 but @KurtE tend to compare notes.

mjs513 commented:
@avolmat-st 

Going to ask is you can do it if you dont mind would be easier - worried I will screw something up :) and right now in middle of some other things.

Sorry.

avolmat-st commented:
I had a deeper look at this point and did some experiment and actually at the end I do not reach exactly the same conclusions.
I actually went a bit fast in my previous message mentioning that the return value given by the DCMI is not correct. It is actually properly returning the right frmival, simply the update of the frmival pointer is done prior to the video_set_frmival to the sensor:



So I did some more check, trying to simulate the experiment you did, using a slighly modified OV9655 sensor driver and the video_shell.
In the OV9655, I forced to have it expose 4 discrete frmival as you have on your hm03b0, that is 60/30/15/10 and also added a print into the set_frmival handler of that driver in order to see what the DCMI driver is requesting to the sensor:

Result is shown below:

```
*** Booting Zephyr OS build v4.3.0-rc3 ***
[00:00:00.312,000] <inf> main: Letting the user control the device with the video shell
uart:~$ video frmival dcmi@48020000
Current frame interval: 33 ms (30 FPS)
			Interval: Stepwise: 66 ms - 16 ms with step 33 ms (60 - 15 FPS)
			Interval: Stepwise: 133 ms - 33 ms with step 66 ms (30 - 8 FPS)
			Interval: Stepwise: 266 ms - 66 ms with step 133 ms (15 - 4 FPS)
			Interval: Stepwise: 400 ms - 100 ms with step 200 ms (10 - 3 FPS)
uart:~$ video frmival dcmi@48020000 20fps
Current frame interval: 33 ms (30 FPS)
New frame interval:     66 ms (15 FPS)
[00:00:34.642,000] <err> video_ov9655: Requested to set num: 1, denominator: 60
```

What this shows is:
1. upon requesting enum_frmival, the dcmi driver will expose the intervals, based on the discrete frmival given by the sensor. The sensor was reporting 60/30/15/10, so the dcmi reports, interval, up to 4 time slower from each discreate frmival from the sensor.
This is I think the expected behavior.

2. When requesting 20fps from the video_shell, the DCMI compute the frmival and chose 15fps which I think is indeed correct in this case. However as we can see in the last line, for this, it is setting the sensor to 60fps and asking to the DCMI to frame drop. This is an issue to be fixed since obviously it should have directly picked the 15fps directly from the sensor instead.

But, the returned framerate given by the video_set_frmival is actually correct, 15fps. This is different from the 60fps given back in your case. I do not understand why in your case you are getting 60fps back from the video_set_frmival actually, could you clarify from where you are getting this value ?

```
Requested FrameRate: 20
Returned FrameRate: 60
```

mjs513 commented:
@avolmat-st - @KurtE 
> I am guessing he is using the output from the call: video_get_frmival
> Which it looks like DCMI mucks with the value returned by the sensor.

the answer to your question is yes I am calling get_frmival:
```
bool Camera::setFrameRate(uint32_t framerate) {
	struct video_frmival frmival;
	int ret;
	
	frmival.numerator = 1;
	frmival.denominator = framerate;
	if (video_set_frmival(this->vdev, &frmival)){
		printk("ERROR: Unable to set up frame rate\n") ;
		return false;
	}
	
	ret = video_get_frmival(this->vdev, &frmival);
	if(ret < 0) Serial.println("ERROR getting framerate...");
	Serial.print("Requested FrameRate: "); Serial.println(framerate);
	Serial.print("Returned FrameRate: "); Serial.println(frmival.denominator);

  return true;
}
```
As @KurtE mentioned I tend to used the ArduinoCore-zephyr while Kurt uses zephyr direct - we cover both bases.  Either way I am printing out the denominator.  But regardless of this form, as I mentioned if I look at debug code from the set_frmival in the 360 driver I see:
```
_set_frmival: FrameRate selected: 60 
hm0360_set_frmival: Applying frame interval number 0
hm0360_set_frmival: OSC DIV 60: 0x1
 HIRES Selected: 0
FrameRate rounded to: 60
```
that set_frmival thinks the that the framerate I passed to it is 60 even though what I passed was 20.  Which is the real problem.  Howver I pass frmival 10, 15, or 30 there is no problem

mjs513 commented:
@avolmat-st - @KurtE 
Just incorporated your changes to DCMI and seems like it resolved the issue>

1.  Input 20 rounded to 15.
2. Input 40 rounded to 30.
 So looks like that resolved the issues.

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -239,32 +415,221 @@ static int hm01b0_soft_reset(const struct device *dev)
+
+	for (;;) {
+		ret = video_read_cci_reg(&config->i2c, HM01B0_CCI_ID, &model_id);
+		if (ret >= 0) {
```
I think success case is simply == 0 here, not >= 0

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -239,32 +415,221 @@ static int hm01b0_soft_reset(const struct device *dev)
+			return false;
+		}
+
+		LOG_INF("Error reading id reg (%d)", ret);
```
How expected is it to get a failing ID register ? If this is quite expected to get 1 or 2 retry, then LOG_DBG might be better.

Moreover, if sometimes retry is needed, isn't this simply that delay post reset / powerdown control is not enought  (CHECK_CONNECTION_DELAY_MS ??)

KurtE replied:
When this code was put in (other platforms), I noticed that several other drivers were doing similar things.
I was getting some failures. before I added a delay after I set the reset/powerdown registers, but if those
are not defined, I don't put in a delay as who ever called us may have done it before calling us.  
And so 1 repeat should always be sufficient.

KurtE commented on drivers/video/CMakeLists.txt:
```diff
@@ -28,6 +28,6 @@ zephyr_library_sources_ifdef(CONFIG_VIDEO_ST_MIPID02	video_st_mipid02.c)
 zephyr_library_sources_ifdef(CONFIG_VIDEO_RENESAS_RA_CEU video_renesas_ra_ceu.c)
 zephyr_library_sources_ifdef(CONFIG_VIDEO_STM32_JPEG	video_stm32_jpeg.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_HIMAX_HM01B0  hm01b0.c)
+zephyr_library_sources_ifdef(CONFIG_VIDEO_HM01B0  hm01b0.c)
```
Where does one do it?

This file?
https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/migration-guide-4.4.rst

Only 1 change in it so far and that was to put general outline in it.  The 4.3 version also looks like just one change.

KurtE replied:
Sorry I guessing as I mentioned that file needs to change... Tried adding:
```
Video
*****

* hm01b0

  * CONFIG_VIDEO_HIMAX_HM01B0
    renamed to CONFIG_VIDEO_HM01B0

Other subsystems
```
With about 5 variations in formats, but the CI tests have tons if issues.
Not sure of the format it is expecting. 

I have not pushed that up yet.

KurtE commented on drivers/video/hm01b0.c:
```diff
@@ -239,32 +415,221 @@ static int hm01b0_soft_reset(const struct device *dev)
+{
+	struct hm01b0_data *drv_data = dev->data;
+
+	*frmival = drv_data->cur_frmival;
```
I now have it initialize it to compute from what the register is set to. 
I currently save it away in the data structure.  The other option is to only compute it if
the user ever asks for it.

KurtE committed:
drivers: video: Update migration-guide

Rename the config CONFIG_VIDEO_HIMAX_HM01B0
to CONFIG_VIDEO_HM01B0

Signed-off-by: Kurt Eckhardt <kurte@rockisland.com>
Co-Authored-By: Alain Volmat <58030053+avolmat-st@users.noreply.github.com>

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -263,8 +586,40 @@ static bool hm01b0_check_connection(const struct device *dev)
+#endif
+#endif
+	/* give time for the camera to startup before checking it */
+	k_msleep(CHECK_CONNECTION_DELAY_MS);
```
Just a guess but, I feel like this extra delay is only meaningful if there there is either a reset or powerdown gpio involved, otherwise it is adding an extra delay on a device which has already been in that state for already a while probably, in which case it should be done within the 
```c
#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(pwdn_gpios) || DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
```

avolmat-st reviewed (changes requested):
Thanks. Looks good to me. Only had few last very minor points.
Also could you rebase on latest head since currently the branch is not mergable ?

KurtE commented on drivers/video/hm01b0.c:
```diff
@@ -239,11 +406,167 @@ static int hm01b0_soft_reset(const struct device *dev)
+	const struct hm01b0_config *config = dev->config;
+	struct hm01b0_data *drv_data = dev->data;
+	struct hm01b0_ctrls *ctrls = &drv_data->ctrls;
+
```
done, next push

avolmat-st reviewed (changes requested):
Thanks @KurtE, sorry (again), one last small wording in the migration-guide tp integrate it better in the generated documentation.

avolmat-st reviewed (dismissed):
Thanks a lot @KurtE. LGTM.

avolmat-st commented:
Hi @KurtE,
can you rebase this PR and fix conflicts in migration-guide-4.4.rst file ?

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -263,8 +586,40 @@ static bool hm01b0_check_connection(const struct device *dev)
+		if (!gpio_is_ready_dt(&config->reset)) {
+			return -ENODEV;
+		}
+		ret = gpio_pin_configure_dt(&config->reset, GPIO_OUTPUT_ACTIVE);
```
Same here for the reset pin. It set it to ACTIVE state before using the sensor. I'd have expected to have probably both powerdown and reset pin set to INACTIVE state in order to have the sensor in normal condition of usage. Moreover, is there a reset sequence available ? Such as ensure that the sensor is kept in reset state at least xx usec before release the reset etc ?
As an example, for the imx335 sensor:

```c
#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
        if (cfg->reset_gpio.port != NULL) {
                if (!gpio_is_ready_dt(&cfg->reset_gpio)) {
                        LOG_ERR("%s: device %s is not ready", dev->name,
                                cfg->reset_gpio.port->name);
                        return -ENODEV;
                }

                /* Power up sequence */
                ret = gpio_pin_configure_dt(&cfg->reset_gpio, GPIO_OUTPUT_ACTIVE);
                if (ret) {
                        return ret;
                }

                k_sleep(K_NSEC(500)); /* Tlow */

                gpio_pin_set_dt(&cfg->reset_gpio, 0);

                k_sleep(K_USEC(600)); /* T4 */
        }
#endif
```

You can see above that at the end of the sequence the reset state is at INACTIVE (0). Which doesn't actually mean that the pin is at 0, it can also be 1 if within the device-tree the gpio is written as ACTIVE_LOW.

avolmat-st replied:
Based at what is written in the binding, the reset is mentioned at a ACTIVE_LOW signal (which makes sense since it is named RESETn. In such case, in the device-tree the pin should be written as ACTIVE_LOW (I guess it is written currently as ACTIVE_HIGH would would make the current code work), and the driver must set the reset state at 0 before starting to use the sensor.

josuah replied:
Connecting the threads together just for convenience:

- https://github.com/zephyrproject-rtos/zephyr/issues/98879#issuecomment-3548565967

avolmat-st reviewed (changes requested):
I have some doubt about the power-down / reset handling, could you check ?

Also, thanks, but no need to add me as Co-Authored-by in the commit, but if you really want to, my email address is : alain.volmat@foss.st.com

KurtE commented on dts/bindings/video/himax,hm01b0.yaml:
```diff
@@ -33,6 +33,16 @@ properties:
+    type: phandle-array
+    description: |
+      The PWDN pin is asserted to power down the sensor. The sensor
+      receives this as an active high signal
```
As I mentioned in another response.  I believe you are right, in that the sensor itself may not have direct support
for these pins, but many/most of the carrier boards do.  In particular the ones I mentioned do this by turning
off the power by disabling a voltage regulator, that supplies the power to the camera. Obviously we can change
the wording here if you like.   Suggestion?

avolmat-st replied:
Did you had a chance to have a look at gpio-hog I mentioned yesterday and see it would fit your need ? I think that in current situation / usage of those pins, actually gpio-hog might be enough.

josuah commented on include/zephyr/drivers/video.h:
Could you remove the whole block `/** ... */` from the commit message? This is already the same as the commit content.

josuah reviewed (commented):
Thank you for coming back to it.

Here are a few changes that still might need to happen.

But few of them as has already been reviewed plenty in the past an you are now used to video drivers.

KurtE commented on drivers/video/hm01b0.c:
```diff
@@ -14,83 +15,248 @@
+	{HM01B0_CCI_FS_50HZ_H,		0x0032},
+
+	{HM01B0_CCI_MD_CTRL,		0x00},
+	{HM01B0_CCI_FRAME_LENGTH_LINES, HIMAX_FRAME_LENGTH_QVGA},
```
done next push

josuah commented on drivers/video/hm01b0.c:
```diff
@@ -14,82 +15,232 @@
+	{HM01B0_CCI_FS_CTRL,		0x00},	/* Flicker Control */
+
+	{HM01B0_CCI_FS_60HZ_H,		0x003C},
+	{HM01B0_CCI_QVGA_WIN_EN,	0x01},	/* Enable QVGA window readout */
```
Sorry I forgot to point at this one which is already in the format-specific tables below and can be removed from here.

Might not be critical, just very minor improvement.

josuah reviewed (commented):
I think it is completely acceptable to only solve the problems you wish to solve upstream, and solve other problems in your own way.

I am grateful for the amount of free time spent on this driver.

Some minor changes found in case you wanted to address them. Not approving yet just to let you proofread the commit message but besides, LGTM.

iabdalkader commented:
> Now suppose some cameras have defined that the reset pin set to physically HIGH(1) puts it into a reset state and some cameras define setting the pin into a LOW state (0), how should all of these be defined?

I don't have an answer to this question, which I think you've asked us before somewhere else. IMHO Zephyr, its subsystems, drivers, device trees etc.. was not designed with this in mind. It's great that we've managed to enable so many different use cases at runtime, but maybe you've hit a fundamental limitation. FWIW I've faced the exact same problem a long time ago with a Linux-based camera, and my solution was to have uboot detect the attached sensor, by toggling I/Os and scanning, and then loading a binary device tree blob (`.dtb`) for the detected sensor (I had one for every supported sensor). Don't know if Zephyr can support something similar.

FWIW, Himax aside, all the cameras you're trying to support are obsolete, decade-old chips that aren't that useful in the real world. Their regs are not very-well documented, their datasheets are obscure, to say the least, and more importantly their Supply chain reliability is questionable, hence not very useful outside hobby projects.

josuah commented:
> and then loading a binary device tree blob (.dtb) for the detected sensor (I had one for every supported sensor). Don't know if Zephyr can support something similar.

Thank you for chiming in.

This is long-standing discussion useful to many vendors and faced by hobbyists and devicetree maintainers (in their current position/activity) alike!

Related:

- https://github.com/zephyrproject-rtos/zephyr/pull/96652
- https://github.com/zephyrproject-rtos/zephyr/issues/98879

KurtE commented on drivers/video/hm01b0.c:
```diff
@@ -263,6 +571,7 @@ static bool hm01b0_check_connection(const struct device *dev)
 
 static int hm01b0_init(const struct device *dev)
 {
+	const struct hm01b0_config __maybe_unused *config = dev->config;
```
in next push

KurtE commented on drivers/video/hm01b0.c:
```diff
@@ -14,82 +15,232 @@
 
 struct hm01b0_config {
 	const struct i2c_dt_spec i2c;
-	const uint8_t data_bits;
```
noticed it awhile ago

josuah reviewed (dismissed):
@avolmat-st let me know if there were still things missing from what you could see.

Thank you everyone.

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -14,82 +15,228 @@
+	{HM01B0_CCI_MAX_AGAIN_BIN2,	0x04},	/* Max Analog gain bin2 mode       [Def: 0x04] */
+	{HM01B0_CCI_MAX_DGAIN,		0xC0},
+
+	{HM01B0_CCI_INTEGRATION_H,	0x0108},	/* Integration H           [Def: 0x01] */
```
Not mandatory here, just mentioning since I saw that, since this INTEGRATION register is mentioned as HIGH, I guess that there must be a LOW register as well. Aren't they contiguous to each other ? If so shouldn't this register be a REG16 instead of REG8 ?
Since this also comment application in some other parts of the code, but again, not mandatory for this PR.

avolmat-st reviewed (changes requested):
A minor point regarding unused newly added define and questionning about the pixel format / resolution handling.
(sorry I might have simply forgotten or missread one part of the driver here, could you please clarify it for me ?)

KurtE commented on drivers/video/hm01b0.c:
```diff
@@ -14,82 +15,228 @@
+	[RESOLUTION_324x244_Y4] = hm01b0_320x240_regs,
+	[RESOLUTION_324x324_Y4] = hm01b0_320x320_regs,
+	[RESOLUTION_324x244_BAYER] = hm01b0_320x240_regs,
+	[RESOLUTION_324x324_BAYER] = hm01b0_320x320_regs,
```
The Y4 stuff... 

From the dts\bindings\video\himax,hm01b0.yaml:
```
properties:
  data-bits:
    type: int
    default: 1
    description: |
      Camera output data width. 1/4/8 bits.
      The default output data width value is 1, pixel value is serialised.
    enum: [1, 4, 8]
```
Which is read in during the init setup.

```
#define HM01B0_INIT(inst)                                                                       \
	const struct hm01b0_config hm01b0_config_##inst = {                                     \
		.i2c = I2C_DT_SPEC_INST_GET(inst),                                              \
		.ctrl_val = HM01B0_CTRL_VAL(DT_INST_PROP(inst, data_bits))                      \
	};                                                                                      \
	struct hm01b0_data hm01b0_data_##inst;                                                  \
	DEVICE_DT_INST_DEFINE(inst, &hm01b0_init, NULL, &hm01b0_data_##inst,                    \
			      &hm01b0_config_##inst, POST_KERNEL, CONFIG_VIDEO_INIT_PRIORITY,   \
			      &hm01b0_driver_api);                                              \
	VIDEO_DEVICE_DEFINE(hm01b0_##inst, DEVICE_DT_INST_GET(inst), NULL);

DT_INST_FOREACH_STATUS_OKAY(HM01B0_INIT)
```

So the bits per pixel clock transferred, can be either 1, 4, 8 default 1 (personally should be 8), but after
the last round of changes I have not modified this file.

So if it were what I think is the default of 8 bits transferred, we receive a complete pixel
per pixel clock, so for Width of 324 we receive 324 clocks and bytes transferred.

With the case of setting to 4, each pixel requires two pixel clocks which generates two bytes in our transfer...

And if you go to the current default setting of 1, each pixel requires 8 pixel clocks and as such 8 bytes transferred.  Now in the original PR, he had it also include an RPI? driver that it went through something like a pipeline which he had hard coded to to combine 8 transfers into 1...

So then back in the 4 bit setting, what logically is the pixel format returned by the camera?  That is where the Y4 came in...  There probably should be a Y1 as well, but... 

As for Bayer versus GREY, I am not sure I have anywhere to easily differentiate them, although there was
some stuff in manual, that it does not support QQVGA?

KurtE replied:
@josuah @avolmat-st - Sounds good.  Not sure, should I change these now, or merge this one in first
and then create new PR? 

Thanks again
